runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -85,45 +123,83 @@ public void startBundle() {
     doFnRunner.startBundle();
   }
 
+  @Override
+  public void finishBundle() {
+    doFnRunner.finishBundle();
+  }
+
   @Override
   public void processElement(WindowedValue<InputT> input) {
 
     // StatefulDoFnRunner always observes windows, so we need to explode
     for (WindowedValue<InputT> value : input.explodeWindows()) {
-
       BoundedWindow window = value.getWindows().iterator().next();
-
       if (isLate(window)) {
         // The element is too late for this window.
-        droppedDueToLateness.inc();
-        WindowTracing.debug(
-            "StatefulDoFnRunner.processElement: Dropping element at {}; window:{} "
-                + "since too far behind inputWatermark:{}",
-            input.getTimestamp(),
-            window,
-            cleanupTimer.currentInputWatermarkTime());
+        reportDroppedElement(value, window);
+      } else if (requiresTimeSortedInput) {
+        processElementOrdered(window, value);
       } else {
-        cleanupTimer.setForWindow(value.getValue(), window);
-        doFnRunner.processElement(value);
+        processElementUnordered(window, value);
+      }
+    }
+  }
+
+  private void processElementUnordered(BoundedWindow window, WindowedValue<InputT> value) {
+    cleanupTimer.setForWindow(value.getValue(), window);
+    doFnRunner.processElement(value);
+  }
+
+  private void processElementOrdered(BoundedWindow window, WindowedValue<InputT> value) {
+
+    StateInternals stateInternals = stepContext.stateInternals();
+    TimerInternals timerInternals = stepContext.timerInternals();
+
+    if (!timerInternals.currentInputWatermarkTime().isAfter(value.getTimestamp())) {
+      StateNamespace namespace = StateNamespaces.window(windowCoder, window);
+      BagState<WindowedValue<InputT>> sortBuffer = stateInternals.state(namespace, sortBufferTag);
+      ValueState<Instant> minStampState = stateInternals.state(namespace, sortBufferMinStampTag);
+      sortBuffer.add(value);
+      Instant minStamp =
+          MoreObjects.firstNonNull(minStampState.read(), BoundedWindow.TIMESTAMP_MAX_VALUE);
+      if (value.getTimestamp().isBefore(minStamp)) {
+        minStamp = value.getTimestamp();
+        minStampState.write(minStamp);
+        setupFlushTimerAndWatermarkHold(namespace, minStamp);
       }
+    } else {
+      reportDroppedElement(value, window);
     }
   }
 
   private boolean isLate(BoundedWindow window) {
     Instant gcTime = LateDataUtils.garbageCollectionTime(window, windowingStrategy);
-    Instant inputWM = cleanupTimer.currentInputWatermarkTime();
+    Instant inputWM = stepContext.timerInternals().currentInputWatermarkTime();
     return gcTime.isBefore(inputWM);
   }
 
+  private void reportDroppedElement(WindowedValue<InputT> value, BoundedWindow window) {
+    droppedDueToLateness.inc();
+    WindowTracing.debug(
+        "StatefulDoFnRunner.processElement: Dropping element at {}; window:{} "
+            + "since too far behind inputWatermark:{}",
```

Comment:
Is this message correct in case of processing and dropping from ordered code path?

Reply:
Should be. But we need to modify the flushing behavior to respect allowed lateness. This is wrong in the current implementation. Will fix that.

Reply:
Fixed.